be/tool/beck.c

```diff
@@ -1556,6 +1632,167 @@ static void genadd(uint64_t gen)
 	}
 }
 
+static int nv_scan_offset_init(uint64_t workers_nr,
+			       uint64_t partitions_nr)
+{
+	uint64_t  p;
+
+	M0_PRE(partitions_nr <= AO_NR);
+
+	m0_mutex_init(&off_info.oi_lock);
+	m0_mutex_lock(&off_info.oi_lock);
+	off_info.oi_workers_nr = workers_nr;
+	off_info.oi_partitions_nr = partitions_nr;
+
+	off_info.oi_offset = m0_alloc(sizeof(struct worker_off_info) *
+				      workers_nr);
+	if (off_info.oi_offset == NULL) {
+		m0_mutex_unlock(&off_info.oi_lock);
+		return M0_ERR(-ENOMEM);
+	}
+
+	memset(&off_info.oi_pinfo, 0, sizeof(struct part_info));
+	for (p = 0; p < off_info.oi_partitions_nr; p++)
+		m0_mutex_init(&off_info.oi_part_lock[p]);
+	m0_mutex_unlock(&off_info.oi_lock);
+	return 0;
+}
+
+static void nv_scan_offset_fini(void)
+{
+	uint64_t  p;
+
+	m0_mutex_lock(&off_info.oi_lock);
+	m0_free(off_info.oi_offset);
+	for (p = 0; p < off_info.oi_partitions_nr; p++)
+		m0_mutex_fini(&off_info.oi_part_lock[p]);
+	m0_mutex_unlock(&off_info.oi_lock);
+	m0_mutex_fini(&off_info.oi_lock);
+}
+
+static off_t nv_scan_offset_get(off_t snapshot_size)
+{
+	FILE     *ofptr;
+	int       wret;
+	int       pret;
+	int       sret;
+	off_t     offset = snapshot_size;
+	off_t     max_offset = 0;
+	uint64_t  p;
+	uint64_t  w;
+	struct part_info        *pinfo;
+	struct worker_off_info  *winfo;
+	struct scanner_off_info *sinfo;
+
+	m0_mutex_lock(&off_info.oi_lock);
+	ofptr = fopen(offset_file, "r");
+	if (ofptr == NULL) {
+		m0_mutex_unlock(&off_info.oi_lock);
+		return 0;
+	}
+	wret = fread(off_info.oi_offset, sizeof(struct worker_off_info),
+		     off_info.oi_workers_nr, ofptr);
+	pret = fread(&off_info.oi_pinfo, sizeof(struct part_info),
+		     1, ofptr);
+	sret = fread(&off_info.oi_scanoff, sizeof(struct scanner_off_info),
+		     1, ofptr);
+	if ((wret > 0) && (pret > 0) && (sret > 0)) {
```

Comment:
these check should be exact = number of bytes

Reply:
How we ensure file data are correct? What is the impact if crash happens after 1st or 2nd fwrite operation?

Reply:
> these check should be exact = number of bytes
Yes, will update that.
> How we ensure file data are correct? What is the impact if crash happens after 1st or 2nd fwrite operation?
Ideally we could use checksum, and do single read/write operations. But due to dynamic allocation worker_off_info there are separate buffers. Having constant defined for MAX limit of workers_nr will simplify this, but not sure what would be such limit.

Reply:
defined MAX_WORKERS_NR limit, to do static allocation of workers offset info. Now single fread/write call is used to save non-volatile offset information.
header magic is added for validation, but checksum cann't be added as multiple workers can be updating non-volatile offset information simultaneously at their respective indexes. There are chances of this info getting updated in between checksum calculation to actual write completion time window.